Jenkins Development Setup Fixes

jacderida opened the pull request:
Hi Stephen/Calum,

Making all those changes for the production infrastructure, I didn't really get a chance to go back and check if they were compatible with the dev infrastructure and the local setup. This has been addressed and both of those are working again.

I've switched the dev infrastructure to use Terraform, so we don't need to worry about manually messing around with security groups and whatnot. None of the AWS dev machines are defined using Vagrant now, so that tidies up the Vagrantfile a bit.

You can spin up the local environment with `make env-jenkins-dev-vbox` and the AWS dev environment with `make env-jenkins-dev-aws`. You should see a Linux and a Windows slave in the local setup and 2 Linux and a Windows slave in the AWS dev setup.

Cheers,

Chris

jacderida committed:
refactor/env: describe dev infra with terraform

Switches from using Vagrant machines and ad-hoc scripts to Terraform for
describing the infrastructure for the dev environment. This is much
better because Terraform is idempotent and the Vagrantfile is getting
quite crowded as it is.

For the dev environment there are 4 machines:

* Jenkins Master
* Docker Slave 1
* Docker Slave 2
* Windows Slave

The Mac is still out of action until we get the VPN working.

jacderida committed:
refactor/env: change dev environment process

jacderida committed:
refactor/vm: get password before provision

Changes the Windows machine provisioning for the dev environment to
retrieve the password at the same time it's going to run Ansible. By
this time the password should be available on the instance, so the loop
won't really have much effect. It was previously used when the
password was being retrieved after the instance was spun up.

Since we aren't using Vagrant for this setup any more, we need to
retrieve the hostname of the Jenkins master via the awscli.

jacderida committed:
refactor/env: change master and slave configs

Changes the docker-compose definition and the configured slaves based on
the type of environment we're running.

In the local and the AWS dev setups, we don't need WireGuard installed,
so this needs to be removed from the compose definition. It actually
doesn't work in these setups (although I'm not sure why), so it had to
be removed, otherwise we could have just left it in and it wouldn't have
done too much harm.

The dev environment will just use a couple of static Linux slaves, so we
remove the use of the EC2 plugin and loop over the Linux slaves group,
as it was before.

jacderida committed:
docs/env: remove vagrant instructions

Vagrant has been removed from the AWS setup entirely.

There is also a little change here to use the master repository for the
production provision, since the provisioning code has been merged into
the main repo.

jacderida committed:
docs/setup: advise using a virtualenv

This is probably a better solution for all the Python libraries.
Sometimes these installations can conflict with system libraries and for
that reason they either won't install, or you can override the system
library and that breaks other stuff.

jacderida committed:
style/targets: reorganise makefile

This reorganises the targets in the Makefile into 5 categories:

* Targets prefixed with box build Packer boxes
* Targets prefixed with vm produce VMs
* Targets prefixed with env produce a full environment (multiple VMs)
* Targets prefixed with provision run Ansible against machines
* Targets prefixed with clean removing environments or VMs

Since we can target stuff on either VirtualBox or AWS, the targets are
often suffixed with either of these.

This commit also removes WireGuard sandbox machines, since it's not
really appropriate to have that sandbox in this repository.

jacderida committed:
docs/targets: update target references

jacderida committed:
docs/setup: macOS workaround

Advise the user to set this variable in their ~/.bashrc, as it doesn't
seem to work reliably when it's set in the Makefile.

jacderida committed:
docs/targets: update more target refs

S-Coyle committed:
fix/vagrantfile: disable synced folders for Windows machine creation

S-Coyle reviewed (dismissed):
Was able to follow setup instructions for Local Provisioning and get SCL Jenkins pipeline up & running on Windows & Linux slaves.
👍

jacderida committed:
Merge branch 'dev_setup' of github.com:jacderida/safe-build-infrastructure into dev_setup

jacderida committed:
refactor/env: remove manual steps

Removes the need to set the Jenkins master location and the Windows
password manually before the production provision.

jacderida committed:
fix/vm: forward wg0 to eth0

This seems to enable the internet access required for the Mac. It's not
completetly clear that this will work with Docker.

jacderida committed:
refactor/role: change role name

The name `ansible-provisioner` is more appropriate for this role,
because it's doing a lot more than just installing Ansible.

jacderida committed:
refactor/env: remove setting subnet

This can be retrieved with the CLI, so we can now remove this step.

jacderida committed:
feat/role: install awscli and jq

These are used in scripts to retrieve the Windows password, the Jenkins
master URL and the ID of the private subnet.

jacderida committed:
fix/script: retrieve subnet id

jacderida committed:
docs/env: set region for prod provision

It's simpler to set this rather than tell them to run `aws configure`.

It may perhaps be better to get the user to provide these on the
environment creation.

jacderida committed:
refactor/env: improve prod env process

I've been working with this setup enough such that the manual steps were
slowing me down, due both to errors in the steps and because they take a
bit of time.

The beginning of the process will prompt you for your AWS keys and the
vault password, and these will then be setup on the Bastion host. It
will also copy the `jenkins_env` key to the Bastion, as this is required
to get the password for the Windows instance.

It should be noted that when we provision the infrastructure for real,
we should remove all of this private information from the Bastion host.
When we can get the proper infrastructure setup, we can do an SSH tunnel
to avoid having to put any keys on the Bastion.

jacderida committed:
feat/env: open port 51820 for wireguard

jacderida committed:
fix/role: remove private key provision

It seemed to be difficult to read the contents of the key into a
variable in the Makefile, so I'm just using a simple script to SCP the
file over.

jacderida committed:
fix/script: replace wget with curl

For some completetly bizarre reason, on the AWS Windows machine, wget
stopped working. It was complaining about a low level library not being
available (despite it working perfectly fine previously).

Instead we can just use curl to do the equivalent thing.

jacderida committed:
fix/script: correct command format

jacderida committed:
fix/script: use {} in variable reference

Without having these it can't seem to correctly parse the variables
around the colon.

jacderida committed:
fix/role: re-enable mingw package

Whatever was wrong with this upstream now seems to have been fixed.

jacderida committed:
feat/env: change to 1 executor per machine

For the machines we're using, they're too small to run more than one job
on their own - it's just too slow. We would need to consider adding
extra machines or using machines with more resources.

jacderida committed:
docs/env: provision macOS slave

Describes how to provision the macOS slave and advises reconnecting the
slave if it goes offline. There are also changes to the other sections
on the production infrastructure just to make things a bit clearer to
understand.

jacderida committed:
fix/vm: use later version of container

jacderida committed:
fix/env: change to medium for windows slave

I found the build was pretty slow when using a t3.small instance, and
it's especially slow if you allow 2 executors to run on the machine.

jacderida committed:
fix/box: update group in box template

jacderida committed:
fix/vm: update linux slave to new ami

jacderida committed:
refactor/env: only use macos slave in prod

The dev environment doesn't have the VPN setup just now, so I'm going to
disable the macOS slave for this environment. Things will need to be
refactored a bit if we want different environments to use the same
machine.

jacderida committed:
fix/env: remove macos provision from dev env

The macOS slave is no longer referenced in the AWS dev setup because the
dev environment isn't running WireGuard.

jacderida committed:
docs/env: provide default region for awscli

jacderida committed:
Merge branch 'vpn_connectivity' into dev_setup

jacderida commented:
Hi Stephen/Calum,

As discussed, I've merged the VPN stuff into this branch.

If one of you could do the dev environment and one of you could do the production environment that would be great. Please read the updated instructions for the relevant environment.

**Important note**: right now you can't provision both environments at the same time. They could be used independently, it's just because of the way Ansible dynamic inventory works we can provision them both at the same time. I will look into this in another branch.

New acceptance criteria:

* Dev environment: environment provisions without any errors
* Dev environment: 2 docker slaves online in Jenkins
* Dev environment: Windows slave online in Jenkins
* Prod environment: creating infrastructure should run without any errors
* Prod environment: provisioning the infrastructure should run without any errors
* Prod environment: Windows slave online in Jenkins
* Prod environment: macOS slave online in Jenkins
* Prod environment: the macOS jobs on the first stage of the SCL pipeline should pass and be able to upload their artifacts to S3

Cheers,

Chris

S-Coyle committed:
refactor/makefile: retrieve required variables from host machine

jacderida committed:
docs/aws: single setup section

Both the development and the production environments really require the
same host setup prerequisites, so I've put this into its own section.
The other sections just describe how to provision the environments.

jacderida committed:
Merge branch 'dev_setup' of github.com:jacderida/safe-build-infrastructure into dev_setup

jacderida committed:
docs/aws: clarify instruction location

This makes it clearer that you need to provision the macOS slave from
your development machine rather than the Bastion host. I also added the
instruction for setting up the vault password.

jacderida committed:
fix/env: remove wireguard for local env

This is not required in the local setup, so the group vars can just
be removed entirely.

S-Coyle committed:
docs/readme: update & clarify various readme instructions

S-Coyle committed:
docs/readme: update & clarify various readme instructions

S-Coyle reviewed (approved):
👍 
Tested prod & dev setup & teardown - after a few changes this is all working well.
As discussed with @jacderida there is an issue with local setup at the moment, but this isn't related to this PR and can be fixed in a future PR.
